Formal2024/Exercises/ex5.lean

```diff
@@ -0,0 +1,102 @@
+import Library
+
+/-! # Part 1 : myEven and myOdd
+  definitions of myEven and myOdd natural numbers using `inductive` -/
+
+inductive myEven : Nat → Prop
+| zero : myEven 0
+| two_plus_n : ∀ n, myEven n → myEven (n + 2)
+
+inductive myOdd : Nat → Prop
+| one : myOdd 1
+| two_plus_n : ∀ n, myOdd n → myOdd (n + 2)
+
+/-! ## basic properties of myEven and myOdd -/
+
+theorem myEven_or_myOdd : ∀ n, myEven n ∨ myOdd n := sorry
+
+theorem myOdd_succ_myEven : ∀ n, myOdd n → myEven (n + 1) := sorry
+
+theorem myEven_succ_myOdd : ∀ n, myEven n → myOdd (n + 1) := sorry
+
+/-! ### addition -/
+
+theorem myEven_add_myOdd : ∀ m n, myEven m → myOdd n → myOdd (m + n) := sorry
+
+theorem myOdd_add_myEven : ∀ m n, myOdd m → myEven n → myOdd (m + n) := sorry
+
+theorem myOdd_add_myOdd : ∀ m n, myOdd m → myOdd n → myEven (m + n) := sorry
+
+theorem myEven_add_myEven : ∀ m n, myEven m → myEven n → myEven (m + n) := sorry
+
+/-! ### multiplication -/
+
+theorem myEven_mul : ∀ m n, myEven m → myEven (m * n) := sorry
+
+theorem mul_myEven : ∀ m n, myEven n → myEven (m * n) := sorry
+
+theorem myOdd_mul_myOdd : ∀ m n, myOdd m → myOdd n → myOdd (m * n) := sorry
+
+/-! ### power -/
+
+theorem myEven_pow : ∀ m n, myEven m → myEven (m ^ n) := sorry
+
+theorem myOdd_pow : ∀ m n, myOdd m → myOdd (m ^ n) := sorry
+
+/-! # Part 2 : uniqueness of natural numbers with `≤`
+
+  `∃! x` is notation for "there exists a unique `x`".
+  This is equivalent to `∃ x, p x ∧ ∀ y, p y → y = x`.
+
+  This can be confusing when involving more variables, so we will prove the following.
+-/
+
+theorem Nat.exists_unique_le : ∃! n m : Nat, m ≤ n := sorry
+
+/-! # Part 3 : existence implies everything (if and only if the universe is nonempty)
+  in other words, if we know something exists, then everything exists -/
+
+theorem existence_implies_everything_iff {α : Type*} (r : α → Prop) :
+  (∃ x, (r x → ∀ y, r y)) ↔ Nonempty α := sorry
```

Comment:
Loved this one! Is there a way of doing it without choice? 
(Also, I'm slightly confused by the wording before, why `if we know something exists, then everything exists`?)

Reply:
Since it depends on the type being nonempty, I think choice is required.

I mean this particular proof can avoid it by using [`forall_imp_iff_exists_imp`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Basic.html#forall_imp_iff_exists_imp), but that uses choice. So I don't think there is a way around it.

Reply:
Yeah, I don't think that makes much sense, my bad. I think I meant to write "if we know something exists, then if it satisfies a condition `r`, then everything satisfies `r`."

I think this is ["drinker's principle/paradox"](https://en.wikipedia.org/wiki/Drinker_paradox): if someone is in a bar, and that someone is drinking, then everyone in the bar is drinking.

Reply:
[wiki/drinker_paradox#non-empty_domain](https://en.wikipedia.org/wiki/Drinker_paradox#Non-empty_domain) actually states the exact theorem I wrote!!

Reply:
I think by 'choice' you guys mean 'classical logic'? The axiom of choice is definitely not needed for this.

Reply:
Yeah lol

Reply:
> I think by 'choice' you guys mean 'classical logic'? The axiom of choice is definitely not needed for this.

I used the axiom of choice to get an element out of Nonempty, but I was wondering if you could do it another way

Reply:
If `h : Nonempty a`, then `rcases h with ⟨x⟩` will give you `x : a`.

Reply:
That's using choice. It's the same as `let x := some a`.

Reply:
`rcases h with ⟨x⟩` doesn't even work actually 🤔

Reply:
> If `h : Nonempty a`, then `rcases h with ⟨x⟩` will give you `x : a`.

But being allowed to do this implies choice (which is just a function Nonempty a --> a
Weirdly enough, I was able to use `rcases` in the proof at that exercise , but it gives an error when I want to use it to prove choice

Reply:
```lean
lemma thmm {α : Type*} (r : α → Prop) :
  (∃ x, (r x → ∀ y, r y)) ↔ Nonempty α := by
  constructor
  · rintro ⟨x, _⟩
    use x
  · intro h
    rcases h with ⟨x⟩
    by_cases h1 : ∃ (x : α), ¬r x
    · rcases h1 with ⟨y, hy⟩
      use y
      intro hr
      exfalso
      exact hy hr
    · push_neg at h1
      use x
      intro _
      assumption
```

Reply:
```lean
lemma weird {α : Type*} : Nonempty α → α := by
  intro h
  rcases h with ⟨x⟩ 
  exact x
``` 

This one gives an error

Reply:
`tactic 'cases' failed, nested error:
tactic 'induction' failed, recursor 'Nonempty.casesOn' can only eliminate into Prop`